contracts/IdleCDOCardManager.sol

```diff
@@ -37,56 +40,78 @@ contract IdleCDOCardManager is ERC721Enumerable {
+    uint256 _riskPos2,
+    uint256 _amountPos2
+  ) external returns (uint256) {
+    require(_amountPos1 > 0 || _amountPos2 > 0, "cannot mint with no amount");
 
-    IdleCDOCard _card = new IdleCDOCard(_idleCDOAddress);
-    IERC20Detailed underlying  = IERC20Detailed(IdleCDO(_idleCDOAddress).token());
+    // mint the Idle CDO card
+    _tokenIds.increment();
+    uint256 tokenId = _tokenIds.current();
+    _mint(msg.sender, tokenId);
 
-    // transfer amount to cards protocol
-    underlying.safeTransferFrom(msg.sender, address(this), _amount);
+    IdleCDOCard _card = new IdleCDOCard();
 
-    // approve the amount to be spend on cdos tranches
-    underlying.approve(address(_card), _amount);
+    if (_amountPos1 > 0) {
+      // deposit position 1
+      _depositToCard(_card, _idleCDOPos1Address, _riskPos1, _amountPos1);
 
-    // calculate the amount to deposit in BB
-    // proportional to risk
-    uint256 depositBB = percentage(_risk, _amount);
+      Card memory newCard;
+      newCard.exposure = _riskPos1;
+      newCard.amount = _amountPos1;
+      newCard.cardAddress = address(_card);
+      newCard.idleCDOAddress = _idleCDOPos1Address;
+      _cardMap[_cardIds.current()] = newCard;
 
-    // calculate the amount to deposit in AA
-    // inversely proportional to risk
-    uint256 depositAA = _amount.sub(depositBB);
+      _cards[tokenId].push(_cardIds.current());
+      _cardIds.increment();
+    }
 
-    _card.mint(depositAA, depositBB);
+    if (_amountPos2 > 0) {
+      // deposit position 2
+      _depositToCard(_card, _idleCDOPos2Address, _riskPos2, _amountPos2);
 
-    // mint the Idle CDO card
-    uint256 tokenId = _mint();
-    _cards[tokenId] = Card(_risk, _amount, address(_card), _idleCDOAddress);
+      Card memory newCard;
+      newCard.exposure = _riskPos2;
+      newCard.amount = _amountPos2;
+      newCard.cardAddress = address(_card);
+      newCard.idleCDOAddress = _idleCDOPos2Address;
+      _cardMap[_cardIds.current()] = newCard;
 
-    return tokenId;
-  }
+      _cards[tokenId].push(_cardIds.current());
+      _cardIds.increment();
+    }
 
-  function card(uint256 _tokenId) public view returns (Card memory) {
-    return _cards[_tokenId];
+    return tokenId;
   }
 
-  function burn(uint256 _tokenId) public returns (uint256 toRedeem) {
-    require(msg.sender == ownerOf(_tokenId), "burn of risk card that is not own");
+  function burn(uint256 _tokenId) external {
+    require(msg.sender == ownerOf(_tokenId), "burn of card that is not own");
 
     _burn(_tokenId);
 
-    Card memory pos = card(_tokenId);
-    IdleCDOCard _card = IdleCDOCard(pos.cardAddress);
-    toRedeem = _card.burn();
+    address cardAddress = card(_tokenId, 0).cardAddress;
 
-    // transfer to card owner
-    IERC20Detailed underlying  = IERC20Detailed(IdleCDO(pos.idleCDOAddress).token());
-    underlying.safeTransfer(msg.sender, toRedeem);
+    // withdraw all positions
+    for (uint256 i = 0; i < _cards[_tokenId].length; i++) {
```

Comment:
you should read _cards[_tokenId].length once into a variable

Reply:
done